ddsmt.py

```diff
@@ -262,30 +264,32 @@ def _substitute (subst_fun, substlist, superset, randomized,
     assert (g_smtformula)
     assert (substlist in (g_smtformula.subst_scopes, g_smtformula.subst_cmds,
                           g_smtformula.subst_nodes))
+    if g_args.thorough:
+        min_gran = 1
+    else:
+        min_gran = max(1, len(superset) // 16)
     nsubst_total = 0
-    gran = len(superset)
+    s = deque(superset) 
+    gran = (len(s) + 1) // 2
     
-    while gran > 0:
-        start_time = time.time()
-        if randomized:
-            subsets = [random.sample(superset, gran) for s in range(
-                       0, len(superset), gran)]
-        else:
-            subsets = [superset[s:s+gran] for s in range (
-                       0, len(superset), gran)]
-        tests_performed = 0
-        for subset in subsets:
-            if g_args.roundtime:
-                if time.time() - start_time > g_args.roundtime:
-                    _log (2, "[!!] test round timeout: reducing granularity")
-                    break
-            tests_performed += 1
+    while gran >= min_gran:
+        for i in range ((len(s) + gran - 1) // gran):
+                         #this is ceil(len(s) / gran)
```

Comment:
Why not use `math.ceil` then?

Also, I do have several issues with this code.

`s` is modified below, but in Python, range() is only evaluated once. So, in essence, you loop over a list that may get shorter with each iteration. You prevent problems originating from that by not accessing items by index but with `popleft`, but overall this makes the code rather hard to read. As good practice, ideally you want to avoid introducing possible "traps". And this I would consider a possible trap since problems will arise if someone is not aware and modifies the code, let's say by changing it to not use popleft but accessing elements via its indices.

I suggest the following.
Do NOT modify s, but collect all substituted items in a new list, say, `substitems`, and only after you're done processing `s`, `s = substitems`.

I recommend this:
```
substitems =  []
nsubsets = math.ceil(len(s) / gran)  # maximum possible number of subsets
for i in range(nsubsets):
    ...
    # process s without modifying it
    # add substituted items to substitems

# after the loop set s = substitems
```
Using a variable for nsubsets mostly serves documentation purposes, you can also use `math.ceil(len(s) / gran` in the range call, but then you should document what it is (the maximum possible number of subsets, right?).